<a href="https://colab.research.google.com/github/pedroregato/Alura-MachineLearning/blob/main/Aula5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [16]:
import pandas as pd

uri = "https://gist.githubusercontent.com/guilhermesilveira/4d1d4a16ccbf6ea4e0a64a38a24ec884/raw/afd05cb0c796d18f3f5a6537053ded308ba94bf7/car-prices.csv"
dados = pd.read_csv(uri)
dados.head()

Unnamed: 0  mileage_per_year  model_year      price sold
0           0             21801        2000   30941.02  yes
1           1              7843        1998   40557.96  yes
2           2              7109        2006   89627.50   no
3           3             26823        2015   95276.14   no
4           4              7935        2014  117384.68  yes

In [17]:
a_renomear = { 
    'mileage_per_year' : 'milhas_por_ano',
    'model_year' : 'ano_do_modelo',
    'price' : 'preco',
    'sold' :'vendido'
}
dados = dados.rename(columns=a_renomear)
dados.head()

Unnamed: 0  milhas_por_ano  ano_do_modelo      preco vendido
0           0           21801           2000   30941.02     yes
1           1            7843           1998   40557.96     yes
2           2            7109           2006   89627.50      no
3           3           26823           2015   95276.14      no
4           4            7935           2014  117384.68     yes

In [18]:
a_trocar = {
    'no': 0,
    'yes': 1
}
dados.vendido = dados.vendido.map(a_trocar)
dados.head()

Unnamed: 0  milhas_por_ano  ano_do_modelo      preco  vendido
0           0           21801           2000   30941.02        1
1           1            7843           1998   40557.96        1
2           2            7109           2006   89627.50        0
3           3           26823           2015   95276.14        0
4           4            7935           2014  117384.68        1

In [19]:
from datetime import datetime

ano_atual = datetime.today().year
dados['idade_do_modelo'] = ano_atual - dados.ano_do_modelo
dados.head()

Unnamed: 0  milhas_por_ano  ...  vendido  idade_do_modelo
0           0           21801  ...        1               21
1           1            7843  ...        1               23
2           2            7109  ...        0               15
3           3           26823  ...        0                6
4           4            7935  ...        1                7

[5 rows x 6 columns]

In [20]:
dados ['km_por_ano'] = dados.milhas_por_ano * 1.60934
dados.head()

Unnamed: 0  milhas_por_ano  ...  idade_do_modelo   km_por_ano
0           0           21801  ...               21  35085.22134
1           1            7843  ...               23  12622.05362
2           2            7109  ...               15  11440.79806
3           3           26823  ...                6  43167.32682
4           4            7935  ...                7  12770.11290

[5 rows x 7 columns]

In [21]:
dados = dados.drop(columns = ["Unnamed: 0", "milhas_por_ano", "ano_do_modelo"], axis=1)
dados.head()

preco  vendido  idade_do_modelo   km_por_ano
0   30941.02        1               21  35085.22134
1   40557.96        1               23  12622.05362
2   89627.50        0               15  11440.79806
3   95276.14        0                6  43167.32682
4  117384.68        1                7  12770.11290

In [22]:
x = dados[["preco", "idade_do_modelo", "km_por_ano"]]
y = dados["vendido"]

In [23]:
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.svm import LinearSVC
from sklearn.metrics import accuracy_score

x = dados[["preco", "idade_do_modelo", "km_por_ano"]]
y = dados["vendido"]

SEED = 5
np.random.seed(SEED)
treino_x, teste_x, treino_y, teste_y = train_test_split(x, y, test_size = 0.25,
                                                         stratify = y)
print("Treinaremos com %d elementos e testaremos com %d elementos" % (len(treino_x), len(teste_x)))

modelo = LinearSVC()
modelo.fit(treino_x, treino_y)
previsoes = modelo.predict(teste_x)

acuracia = accuracy_score(teste_y, previsoes) * 100
print("A acurácia foi %.2f%%" % acuracia)

Treinaremos com 7500 elementos e testaremos com 2500 elementos
A acurácia foi 62.80%


/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:947: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


Em nosso projeto de venda de carros, executamos o LinearSVC para sabermos qual seria a porcentagem de acurácia, e tivemos um resultado de 57.88%.

Para definirmos se esse resultado é bom ou ruim, ainda precisamos de uma base comparativa, estipulando qual seria o mínimo de acerto aceitável - por exemplo, se todos os palpites fossem 0 ou 1.

Já fizemos esse procedimento anteriormente por meio de np.ones(), criando um array de 1 com o número de elementos que queríamos utilizar. Como esse é um processo bastante recorrente, as bibliotecas possuem ferramentas que facilitam nosso trabalho.

Como podemos verificar na documentação da biblioteca SKLearn, ele possui um estimador conhecido como DummyClassifier, que é basicamente o que o nome diz: um classificador "bobo".

Isso porque ele possui várias estratégias bem simples de classificação, e que podem ser passadas como parâmetro: '

class sklearn.dummy.DummyClassifier(strategy=’stratified’, random_state=None, constant=None)[source]

Para nossos testes, primeiramente importaremos esse classificador e criaremos a variável dummy como padrão. Em seguida, trabalharemos com os métodos tradicionais fit() e predict(), exatamente como no LinearSVC.

Depois do treino, o DummyClassifier() irá prever os resultados e imprimiremos sua acurácia na tela:

In [24]:
from sklearn.dummy import DummyClassifier

dummy = DummyClassifier()
dummy.fit(treino_x, treino_y)
previsoes = dummy.predict(teste_x)

acuracia = accuracy_score(teste_y, previsoes) * 100
print("A acurácia do dummy foi %.2f%%" % acuracia)

A acurácia do dummy foi 52.44%


/usr/local/lib/python3.7/dist-packages/sklearn/dummy.py:132: FutureWarning: The default value of strategy will change from stratified to prior in 0.24.
  "stratified to prior in 0.24.", FutureWarning)


Tivemos um resultado de 52.44%. Nosso algoritmo LinearSVC fez 10% a mais (57.88%) - 5% em números absolutos, mas 10% em termos relativos.

Portanto, existe uma melhora do LinearSVC em relação do DummyClassifier. Contudo, a estratégia utilizada pelo DummyClassifier é estratificada ou (stratified), ou seja, proporcional; se temos 30% de itens da classe 0 e 70% da classe 1, o estimador manterá esse padrão em sua classificação.

Deixaremos explícito em nosso código que é essa a estratégia utilizada modificando dummy para dummy_stratified e executaremos novamente o código.

In [25]:
from sklearn.dummy import DummyClassifier

dummy_stratified = DummyClassifier()
dummy.fit(treino_x, treino_y)
previsoes = dummy.predict(teste_x)

acuracia = accuracy_score(teste_y, previsoes) * 100
print("A acurácia do dummy foi %.2f%%" % acuracia)

A acurácia do dummy foi 50.68%


/usr/local/lib/python3.7/dist-packages/sklearn/dummy.py:132: FutureWarning: The default value of strategy will change from stratified to prior in 0.24.
  "stratified to prior in 0.24.", FutureWarning)


Teremos um resultado inferior: 50.58%. Isso se deve pelo uso de valores aleatórios (random_state). Como esse estimador também usa o Numpy como base do valor de SEED, se executarmos o código desde o começo, a acurácia voltará a ser 52.44%.

Por esse motivo é importante, em determinados pontos do projeto, reiniciar o runtime e rodar todo o código novamente. Dessa forma, teremos certeza de que o SEED original foi utilizado e que a ordem de execução foi mantida desde o início.

Testaremos uma outra estratégia do DummyClassifier, o argumento most_frequent, ou "mais frequente".

In [26]:
dummy_mostfrequent = DummyClassifier(strategy="most_frequent")
dummy_mostfrequent.fit(treino_x, treino_y)
previsoes = dummy_mostfrequent.predict(teste_x)

acuracia = accuracy_score(teste_y, previsoes) * 100
print("A acurácia do dummy mostfrequent foi %.2f%%" % acuracia)

A acurácia do dummy mostfrequent foi 58.00%


Como obtemos uma acurácia menor, a estratégia dummy_stratified nos parece mais razoável como baseline, e é ela que queremos superar.

Nosso código está repetitivo em relação ao cálculo da taxa de acurácia, e há formas melhores de gerar esse número. Como este é um procedimento muito comum em classificadores, uma das funções disponíveis nessas bibliotecas é o score, que retorna a acurácia do modelo que estamos testando.

Portanto, ao invés de chamarmos o predict, usaremos o score, que precisará receber teste_x e teste_y como parâmetro para realizar as comparações. Além disso, multiplicaremos o resultado por 100 para obtermos o valor em porcentagem e imprimiremos a acurácia na tela:

In [27]:
dummy_stratified = DummyClassifier()
dummy_stratified.fit(treino_x, treino_y)
acuracia = dummy_stratified.score(teste_x, teste_y) * 100

print("A acurácia do dummy stratified foi %.2f%%" % acuracia)

A acurácia do dummy stratified foi 52.64%


/usr/local/lib/python3.7/dist-packages/sklearn/dummy.py:132: FutureWarning: The default value of strategy will change from stratified to prior in 0.24.
  "stratified to prior in 0.24.", FutureWarning)


Com isso, não precisamos importar o accuracy_score(), reduzindo uma linha do nosso código.

Agora tentaremos detectar padrões não necessariamente lineares para nossos dados. Poderíamos cogitar a ideia de utilizar o SVC, mas é importante lembrar que ele (além de outros estimadores) não pode ser usado diretamente com dados em várias escalas diferentes.

Aprendemos no projeto anterior que precisamos reescalar nossos dados utilizando o StandardScaler() do pré-processamento.

Reaproveitaremos o código do projeto anterior e o revisaremos em busca de adaptações necessárias. Nele, temos o SEED fixo, a quebra dos dados de treino e teste e a separação do treino raw (os dados antes de serem reescalados).

Em seguida, o StandardScaler(), a partir da média, desloca esses dados e divide pelo desvio padrão, criando um intervalo. Com esse treino, os dados crus são transformados em treino_x e teste_x, que então são utilizados no treino e no teste do nosso modelo. Por fim, imprimimos a acurácia:

In [28]:
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

SEED = 5
np.random.seed(SEED)
raw_treino_x, raw_teste_x, treino_y, teste_y = train_test_split(x, y, test_size = 0.25,
                                                         stratify = y)
print("Treinaremos com %d elementos e testaremos com %d elementos" % (len(treino_x), len(teste_x)))

scaler = StandardScaler()
scaler.fit(raw_treino_x)
treino_x = scaler.transform(raw_treino_x)
teste_x = scaler.transform(raw_teste_x)

modelo = SVC()
modelo.fit(treino_x, treino_y)
previsoes = modelo.predict(teste_x)

acuracia = accuracy_score(teste_y, previsoes) * 100
print("A acurácia foi %.2f%%" % acuracia)

Treinaremos com 7500 elementos e testaremos com 2500 elementos
A acurácia foi 77.48%


Com esse resultado, tivemos em ganho muito grande de acurácia em relação à baseline - aproximadamente 25% de ganho absoluto e 48% de ganho relativo.

No entanto, ainda não estamos compreendendo quais critérios o algoritmo usa para definir qual carro será vendido e qual não.

Não é trivial estudarmos a estrutura interna desses algoritmos para entendermos por que estão sendo tomadas determinadas decisões. Algoritmos de classificação como o Support Vector Machine são difíceis de estudar e interpretar, e, por esse motivo, são apelidados de "caixa preta" (ou black box, em inglês).

Mas será que existem classificadores que conseguem exibir com clareza quais são as regras de decisão que são tomadas para chegar na classificação? Na verdade sim, e aprenderemos quais são eles nas próximas aulas.